src/AdoPat/Scopes.cs

```diff
@@ -0,0 +1,162 @@
+            "vso.gallery_manage",
+
+            // Notifications
+            "vso.notification",
+            "vso.notification_write",
+            "vso.notification_manage",
+            "vso.notification_diagnostics",
+
+            // Packaging
+            "vso.packaging",
+            "vso.packaging_write",
+            "vso.packaging_manage",
+
+            // Project and Team
+            "vso.project",
+            "vso.project_write",
+            "vso.project_manage",
+
+            // Release
+            "vso.release",
+            "vso.release_execute",
+            "vso.release_manage",
+
+            // Security
+            "vso.security_manage",
+
+            // Service Connections
+            "vso.serviceendpoint",
+            "vso.serviceendpoint_query",
+            "vso.serviceendpoint_manage",
+
+            // Settings
+            "vso.settings",
+            "vso.settings_write",
+
+            // Symbols
+            "vso.symbols",
+            "vso.symbols_write",
+            "vso.symbols_manage",
+
+            // Task Groups
+            "vso.taskgroups_read",
+            "vso.taskgroups_write",
+            "vso.taskgroups_manage",
+
+            // Team Dashboard
+            "vso.dashboards",
+            "vso.dashboards_manage",
+
+            // Test Management
+            "vso.test",
+            "vso.test_write",
+
+            // Tokens
+            "vso.tokens",
+            "vso.tokenadministration",
+
+            // User Profile
+            "vso.profile",
+            "vso.profile_write",
+
+            // Variable Groups
+            "vso.variablegroups_read",
+            "vso.variablegroups_write",
+            "vso.variablegroups_manage",
+
+            // Wiki
+            "vso.wiki",
+            "vso.wiki_write",
+
+            // Work Items
+            "vso.work",
+            "vso.work_write",
+            "vso.work_full",
+        };
+
+        /// <summary>Validate the given scopes.</summary>
+        /// <param name="scopes">The scopes to validate.</param>
+        /// <returns>The set of invalid scopes present in the given scopes. The
+        /// empty set means no scopes were invalid.</returns>
+        public static ImmutableHashSet<string> Validate(IEnumerable<string> scopes)
+        {
+            return scopes.Except(ValidScopes).ToImmutableHashSet();
+        }
+
+        /// <summary>Normalize scopes by deduplicating and converting to lowercase.</summary>
+        /// <param name="scopes">The scopes to normalize.</param>
+        /// <returns>Normalized scopes in sorted order. This might be the empty set.</returns>
+        public static ImmutableSortedSet<string> Normalize(IEnumerable<string> scopes)
+        {
+            var scopesCount = scopes?.Count() ?? 0;
+            return (scopesCount == 0) ?
+                ImmutableSortedSet.Create<string>() :
+                scopes.Select(scope => scope.ToLower()).ToImmutableSortedSet();

```

Comment:
I feel weird because of so many question marks. I hope following code snip can make it cleanner

```
(scopes ?? Enumerable.Empty<string>()).Select(scope => scope.ToLower().ToImmutableSortedSet());
```
or
```
using Microsoft.VisualStudio.Services.Common;
scopes.AsEmptyIfNull().Select(scope => scope.ToLower().ToImmutableSortedSet());
```